### PR TITLE
Division and Modulo by zero

### LINKED ISSUE #2213: Division and Modulo by zero - Wrong behavior
RedisGraph 2.8.9

## First issue:

The following queries do not behave similarly:

```
GRAPH.QUERY XXX "CREATE (:Person {name: 'p1', age: 1 / 0})"
```
Sets the property value to infinity.

While:

```
GRAPH.QUERY XXX "CREATE (:Person {name: 'p1', age: 1 % 0})"
```
Returns an error ('Division by zero').

---

Note that Neo4j does not allow inserting **integer** properties that are evaluated to INF or NaN:

```
CREATE (:Person {name: 'p3', age: 1 / 0})
```
Neo.ClientError.Statement.ArithmeticError
/ by zero

```
CREATE (:Person {name: 'p3', age: 1 % 0})
```
Neo.ClientError.Statement.ArithmeticError
/ by zero

---
## Second issue:

Floating-point division by zero success:
```
GRAPH.QUERY XXX "CREATE (:Person {name: 'p1', age: 1.0 / 0})"
```

While floating-point modulo by zero fails:
```
GRAPH.QUERY XXX "CREATE (:Person {name: 'p1', age: 1.0 % 0})"
```

---

Note that Neo4j allows both
```
CREATE (:Person {name: 'p3', age: 1.0 / 0})
CREATE (:Person {name: 'p3', age: 1.0 % 0})
```
In the first query, the property value is set to `INF`, and in the second it is set to `NaN`.

### PR DESCRIPTION
**Pull request by nafraf**
Hi, this PR implements validations of input arguments of division and modulo functions to fix #2213

Test were implemented to check the same behavior of neo4j. 
The tests only include modulo and division operations.
I have not created tests creating properties, but if you think that they are required, I can add them.
Thanks

### CODE REVIEW

File: tests/flow/test_function_calls.py [hunk 1]
```diff
@@ -987,4 +1003,27 @@ def test35_min_max(self):
...
+        query = "RETURN 3 / 0"
+        try:
+            actual_result = graph.query(query)
+        except redis.ResponseError as e:
+            self.env.assertContains("Division by zero", str(e))
+        
+        # Validate floating-point dividend division by 0
+        query = "RETURN 1.0 / 0"
+        actual_result = graph.query(query)
+        self.env.assertTrue(math.isinf(actual_result.result_set[0][0]))
```
Comment (swilly22) [inline tests/flow/test_function_calls.py:1019]: Why is there a difference between integer and floating point?
the former yields an error while the latter returns infinity.

Should we check for positive and negative infinity ?
e.g. `-1 / 0.0` and `1 / 0.0` ?

File: tests/flow/test_function_calls.py [hunk 2]
```diff
@@ -227,12 +228,27 @@ def test10_modulo_inputs(self):
...
 
-        # Validate modulo by 0
+        # Validate integer dividend modulo by 0
         query = "RETURN 3 % 0"
         try:
             actual_result = graph.query(query)
         except redis.ResponseError as e:
             self.env.assertContains("Division by zero", str(e))
+        
+        # Validate floating-point dividend modulo by 0
```
Comment (swilly22) [inline tests/flow/test_function_calls.py:238]: Can you please extend the comment to include what RedisGraph returns when evaluating `1.0 % 0` ?
i.e. what would be the output on `RedisCli`

File: src/arithmetic/numeric_funcs/numeric_funcs.c [hunk 3]
```diff
@@ -35,18 +35,23 @@ SIValue AR_MUL(SIValue *argv, int argc, void *private_data) {
 /* returns the division of given values. */
 SIValue AR_DIV(SIValue *argv, int argc, void *private_data) {
 	if(SIValue_IsNull(argv[0]) || SIValue_IsNull(argv[1])) return SI_NullVal();
+
+	// check for division by zero e.g. n / 0
+	if((SI_TYPE(argv[0]) & SI_TYPE(argv[1]) & T_INT64) && SI_GET_NUMERIC(argv[1]) == 0) {
+		Error_DivisionByZero();
+		return SI_NullVal();
+	}
```
Comment (swilly22) [inline src/arithmetic/numeric_funcs/numeric_funcs.c:43]: Why do we emit an error only when both arguments are `integer` ?
shouldn't we emit the same error when dealing with floats ?

File: src/arithmetic/numeric_funcs/numeric_funcs.c [hunk 4]
```diff
@@ -35,18 +35,23 @@ SIValue AR_MUL(SIValue *argv, int argc, void *private_data) {
...
 SIValue AR_MODULO(SIValue *argv, int argc, void *private_data) {
 	if(SIValue_IsNull(argv[0]) || SIValue_IsNull(argv[1])) return SI_NullVal();
 
-	// check for division by zero e.g. n % 0
-	if(SI_GET_NUMERIC(argv[1]) == 0) {
+	// check for modulo by zero e.g. n % 0
+	if((SI_TYPE(argv[0]) & SI_TYPE(argv[1]) & T_INT64) && SI_GET_NUMERIC(argv[1]) == 0) {
 		Error_DivisionByZero();
 		return SI_NullVal();
 	}
```
Comment (swilly22) [inline src/arithmetic/numeric_funcs/numeric_funcs.c:54]: Why do we emit an error only when both arguments are integer ?
shouldn't we emit the same error when dealing with floats ?

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: tests/flow/test_function_calls.py (hunk 2 above)
Comment (nafraf) [reply tests/flow/test_function_calls.py:241]: Ok. Comments added in a3523cc.

File: tests/flow/test_function_calls.py (hunk 1 above)
Comment (nafraf) [reply tests/flow/test_function_calls.py]: Ok. those tests were added in a3523cc. 
Tests with both operands equal to zero were added too.

File: src/arithmetic/numeric_funcs/numeric_funcs.c (hunk 3 above)
Comment (nafraf) [reply src/arithmetic/numeric_funcs/numeric_funcs.c:43]: Hi. I did not find any definition for that. The error is emitted only for integer to get the same behavior of neo4j.

File: tests/flow/test_function_calls.py (hunk 1 above)
Comment (swilly22) [reply tests/flow/test_function_calls.py]: I think the idea is like so:
When performing `x/0` the division is undefined as `x` is being divided by an **absolute** 0.
While `x/0.0` is either `+inf` or `-inf` depending on `x`
As `0.0` isn't necessarily an absolute `0` but a **really small number** who's almost reaching `0`

File: src/arithmetic/numeric_funcs/numeric_funcs.c (hunk 4 above)
Comment (swilly22) [reply src/arithmetic/numeric_funcs/numeric_funcs.c:54]: 👍

File: tests/flow/test_function_calls.py [hunk 5]
```diff
@@ -987,4 +1055,106 @@ def test35_min_max(self):
...
+        query = "RETURN 0 / 0.0"
+        actual_result = graph.query(query)
+        self.env.assertTrue(math.isnan(actual_result.result_set[0][0]))
+
+        # Validate floating-point 0.0 divided by integer 0 
+        # redis-cli output example:
+        # 127.0.0.1:6379> GRAPH.QUERY g "RETURN 0.0 / 0"
+        # 1) 1) "0.0 / 0"
+        # 2) 1) 1) "-nan"
+        query = "RETURN  0.0 / 0"
```
Comment (swilly22) [inline tests/flow/test_function_calls.py:1148]: extra space `RETURN  0`

File: tests/flow/test_function_calls.py [hunk 6]
```diff
@@ -227,12 +228,79 @@ def test10_modulo_inputs(self):
...
+        self.env.assertTrue(math.isnan(actual_result.result_set[0][0]))
+
+        # Validate integer dividend modulo by 0.0
+        # redis-cli output example:
+        # 127.0.0.1:6379> GRAPH.QUERY g "RETURN 1 % 0.0"
+        # 1) 1) "1 % 0.0"
+        # 2) 1) 1) "-nan"
+        query = "RETURN 1 % 0.0"
+        actual_result = graph.query(query)
+        self.env.assertTrue(math.isnan(actual_result.result_set[0][0]))
```
Comment (swilly22) [inline tests/flow/test_function_calls.py:257]: from `man fmod`

```
SPECIAL VALUES
     fmod(±0, y) returns ±0 if y is neither 0 nor NaN.

     fmod(x, y) returns a NaN and raises the "invalid" floating-point exception for x infinite or y zero.

     fmod(x, ±infinity) returns x for x not infinite.
```

**Review by swilly22 — CHANGES_REQUESTED** [review]
almost there

File: tests/flow/test_function_calls.py (hunk 6 above)
Comment (nafraf) [reply tests/flow/test_function_calls.py:257]: I'll add test cases for those special values.
Is there a constant to represent infinite? I could test like this:
```
127.0.0.1:6379> GRAPH.QUERY g "RETURN 0 % ( 1 / 0.0 )"
1) 1) "0 % ( 1 / 0.0 )"
2) 1) 1) "0"
```
```
127.0.0.1:6379> GRAPH.QUERY g "RETURN 0.0 % ( 1 / 0.0 )"
1) 1) "0.0 % ( 1 / 0.0 )"
2) 1) 1) "0"
```

But it depends on division correctness.

File: tests/flow/test_function_calls.py (hunk 5 above)
Comment (nafraf) [reply tests/flow/test_function_calls.py]: Fixed in [1572385](https://github.com/RedisGraph/RedisGraph/pull/2585/commits/1572385051c22c77613807e9987fb3b47e44ec41)

File: tests/flow/test_function_calls.py (hunk 6 above)
Comment (nafraf) [reply tests/flow/test_function_calls.py:257]: Tests were added in [1572385](https://github.com/RedisGraph/RedisGraph/pull/2585/commits/1572385051c22c77613807e9987fb3b47e44ec41)

**Review by swilly22 — APPROVED** [review]

**Comment by AviAvni** [comment]
please resolve conflict

**Comment by nafraf** [comment]
@AviAvni Ok. Conflicts resolved. Thanks

**Pull request merged**